From 24eba6a433c5ef38fbd698c1ef49b48baa25656b Mon Sep 17 00:00:00 2001
From: Brian Osman <brianosman@google.com>
Date: Thu, 21 Mar 2024 15:57:36 +0000
Subject: [PATCH] Don't move PaintRecord in FormattedText draw callback

When drop shadows are enabled, the draw function can be invoked more
than once, resulting in a crash.

Bug: b/41492786
Change-Id: I27f08d2a72ec73e50f025aa98f097dc8fbbb6bbd
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5372808
Reviewed-by: Jean-Philippe Gravel <jpgravel@chromium.org>
Commit-Queue: Brian Osman <brianosman@google.com>
Cr-Commit-Position: refs/heads/main@{#1276265}
---

--- a/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d.cc
+++ b/third_party/blink/renderer/modules/canvas/canvas2d/canvas_rendering_context_2d.cc
@@ -674,7 +674,7 @@
     Draw<OverdrawOp::kNone>(
         [&recording](cc::PaintCanvas* c,
                      const cc::PaintFlags* flags)  // draw lambda
-        { c->drawPicture(std::move(recording)); },
+        { c->drawPicture(recording); },
         [](const SkIRect& rect) { return false; }, bounds,
         CanvasRenderingContext2DState::PaintType::kFillPaintType,
         CanvasRenderingContext2DState::kNoImage,
--- a/third_party/blink/web_tests/fast/formatted-text/formattedtext-api.html
+++ b/third_party/blink/web_tests/fast/formatted-text/formattedtext-api.html
@@ -120,4 +120,17 @@
     "run owned by destroyed iframe");
 }, 'Use after iframe destroyed');
 
+// crbug.com/41492786
+test(function (t) {
+  var context = getCanvas2DContext(t.name);
+  context.shadowBlur = 10;
+  context.shadowColor = "khaki";
+  context.textRendering = "auto";
+  context.globalCompositeOperation = "multiply";
+  context.lineCap = "square";
+
+  const text = FormattedText.format("1");
+  context.drawFormattedText(text, 20, 20);
+}, 'Formatted text with shadows');
+
 </script>
